Misson Control Store: Improve performance

matheusd opened the pull request:
The current implementation of the Misson Control Store [has a one second ticker](https://github.com/lightningnetwork/lnd/blob/d9887f32129b21e40cf88c79512b16dc5d1715fb/routing/missioncontrol_store.go#L284) that trims the DB, ensuring only the latest 1000 entries are kept.

That ticker was added in https://github.com/lightningnetwork/lnd/pull/5515 as a way to improve performance by batching and de-duplicating the updates to the db.

However, its current implementation can be significantly improved by doing a few pointed changes. In particular:

1. There is no need to open a db tx when there is no work.
2. The ticker does not need to run continuously, only when there are items in the queue that need to be persisted.
3. The current implementation duplicates the entire in-memory cache of the store, which is not ideal when there are only a few entries that need to be added

This PR improves the situation iteratively on each commit. A set of benchmarks is added to verify the performance improvement.

This is the final result (using a bbolt MC store):

```
MissionControlStoreFlushing/no_additional_results-7      221µs ±15%       0µs ± 2%   -99.86%  (p=0.000 n=10+10)
MissionControlStoreFlushing/one_additional_result-7      359µs ±40%      44µs ± 8%   -87.87%  (p=0.000 n=10+10)
MissionControlStoreFlushing/ten_additional_results-7     344µs ± 9%     129µs ±29%   -62.61%  (p=0.000 n=9+10)
MissionControlStoreFlushing/100_additional_results-7     987µs ±28%     621µs ±22%   -37.06%  (p=0.000 n=10+10)
MissionControlStoreFlushing/250_additional_results-7    2.02ms ±37%    1.53ms ±41%   -24.62%  (p=0.005 n=10+10)
MissionControlStoreFlushing/500_additional_results-7    2.88ms ±11%    2.84ms ±18%      ~     (p=0.780 n=9+10)

name                                                  old alloc/op   new alloc/op   delta
MissionControlStoreFlushing/no_additional_results-7      139kB ± 0%       0kB       -100.00%  (p=0.000 n=10+10)
MissionControlStoreFlushing/one_additional_result-7      153kB ± 0%      19kB ± 0%   -87.37%  (p=0.000 n=10+10)
MissionControlStoreFlushing/ten_additional_results-7     175kB ± 0%      41kB ± 0%   -76.35%  (p=0.000 n=8+8)
MissionControlStoreFlushing/100_additional_results-7     350kB ± 0%     222kB ± 0%   -36.63%  (p=0.000 n=10+10)
MissionControlStoreFlushing/250_additional_results-7     651kB ± 0%     535kB ± 0%   -17.75%  (p=0.000 n=8+10)
MissionControlStoreFlushing/500_additional_results-7    1.14MB ± 0%    1.04MB ± 0%    -8.57%  (p=0.000 n=10+10)

name                                                  old allocs/op  new allocs/op  delta
MissionControlStoreFlushing/no_additional_results-7      1.06k ± 0%     0.00k       -100.00%  (p=0.000 n=10+10)
MissionControlStoreFlushing/one_additional_result-7      1.15k ± 0%     0.12k ± 0%   -90.02%  (p=0.000 n=10+10)
MissionControlStoreFlushing/ten_additional_results-7     1.49k ± 0%     0.46k ± 0%   -69.36%  (p=0.000 n=10+10)
MissionControlStoreFlushing/100_additional_results-7     4.90k ± 0%     3.87k ± 0%   -21.07%  (p=0.000 n=10+10)
MissionControlStoreFlushing/250_additional_results-7     10.6k ± 0%      9.6k ± 0%    -9.75%  (p=0.000 n=10+10)
MissionControlStoreFlushing/500_additional_results-7     20.1k ± 0%     19.1k ± 0%    -5.15%  (p=0.000 n=9+10)
```

As can be seen, there is a significant improvement in cpu and memory usage when up to 100 additional results are added (i.e. up to 100 txs/second flowing through the node).

The improvement is specially significant when no new results are added to a full MC store, which is relevant for nodes that have a low or only episodic volument of payments (e.g. a mostly end user node or mobile node).

matheusd commented:
Lint issues are fixed, I believe the other CI failures are just flakes.

guggero reviewed (commented):
Thanks for the PR. Did a first round, have a couple of questions.

matheusd commented on routing/missioncontrol_store.go:
```diff
@@ -283,13 +297,17 @@ func (b *missionControlStore) run() {
+				timerChan = time.After(b.flushInterval)
+
+			case <-timerChan:
+				timerChan = nil
```
The ticker waking up _every_ second is the annoying part on mobile. This prevents the runtime from idling this goroutine, increasing CPU usage. Granted, it's not by much, but afaict, based on some profiling, this is the only tight loop that does not depend on a long (minute+) timer or on an external event (channel update, gossip msg, p2p block, etc) happening, so it seemed it could easily be improved with this fix.

matheusd replied:
Android has tight restrictions for background operation (no more than 6s of cpu time per 5min interval before an app is killed), so idling as much as possible is desirable.

guggero replied:
Ah, I see. I think you should add that to the commit message, otherwise the next reviewer might ask the same question.

But, this pattern with the channel is very weird. Can you implement this using `sync.Cond` please? Then we don't need to drain any channels or anything like that at the end.

matheusd replied:
I'm not sure how to use `Cond` in this situation, because Cond doesn't allow selecting on it (which is needed here due to `b.done`). 

I guess I could use a timer instead of chans directly here or switch back to the ticker, but stop it if there's no work if that would make more sense to you (though I find this particular pattern of a nil chan shorter and clearer when the timeouts are small enough that you don't really need to stop the timer).

ellemouton replied:
@matheusd  - I think you can do it as follows:

1) have these members in the `missionControlStore`

```
	queue     *list.List
	queueMx   sync.Mutex
	queueCond *sync.Cond
```

2) initialise `queueCond`:
```
mc.queueCond = sync.NewCond(&mc.queueMx)
```

3) replace all calls to `queueMx` with `queueCond.L`

4) update `AddResult`:

```
func (b *missionControlStore) AddResult(rp *paymentResult) {
	b.queueCond.L.Lock()
	b.queue.PushBack(rp)
	b.queueCond.L.Unlock()

	b.queueCond.Signal()
}
```

5) update `run`: 

```
func (b *missionControlStore) run() {
	b.wg.Add(1)

	go func() {
		defer b.wg.Done()

		timer := time.NewTimer(b.flushInterval)
		timer.Stop()

		for {
			// Wait for the queue to not be empty.
			b.queueCond.L.Lock()
			for b.queue.Front() == nil {
				b.queueCond.Wait()

				select {
				case <-b.done:
					b.queueCond.L.Unlock()

					return
				default:
				}
			}
			b.queueCond.L.Unlock()

			// Restart the timer.
			timer.Reset(b.flushInterval)

			select {
			case <-timer.C:
				if err := b.storeResults(); err != nil {
					log.Errorf("Failed to update mission "+
						"control store: %v", err)
				}

			case <-b.done:
				return
			}
		}
	}()
}
```

6) adjust `stop`:

```
// stop stops the store ticker goroutine.
func (b *missionControlStore) stop() {
	close(b.done)

	b.queueCond.Signal()

	b.wg.Wait()
}
```

matheusd replied:
Personally, I find `Cond` and `Timer` harder to reason about then just standard channels (but that's really on me).

In any case, I updated the code with your suggested implementation. Thank you for this!

matheusd commented on go.mod:
```diff
@@ -4,6 +4,7 @@ require (
 	github.com/NebulousLabs/go-upnp v0.0.0-20180202185039-29b680b06c82
 	github.com/Yawning/aez v0.0.0-20211027044916-e49e68abd344
 	github.com/andybalholm/brotli v1.0.3
+	github.com/bahlo/generic-list-go v0.2.0
```
Fair enough, though I should note that this _specific_ dependency is very small (the generic list impl is the only thing in it), it's a drop-in replacement, it's based on stdlib code, it's complete, fully tested has no open issues and hasn't had a need for any new code in over 2 years, so it's about as good a dependency as anyone could hope for :p

To bring this into the fn package, I'd just copy the subset of the code that was used here over (code is BSD licensed). Would that be ok?

matheusd replied:
What were the changes you did on this benchmark? Because the `no_additional_results` case is significantly different from the original benchmark in the PR:

```
MissionControlStoreFlushing/no_additional_results-7      221µs ±15%       0µs ± 2%   -99.86%  (p=0.000 n=10+10)
```

lightninglabs-deploy commented:
@bitromortac: review reminder
@matheusd, remember to re-request review from reviewers when ready

bitromortac reviewed (commented):
Nice improvements in relative terms. In absolute terms, on my machine this reduces execution time from 30 ms to 30 µs for adding a single result to 100000 preexisting results (a value some people use if they want a more complete picture of the graph). I'm still thinking about the added de-duplication code complexity, perhaps you could give more motivation for the change (maybe benchmarks on weaker hardware)?

lightninglabs-deploy commented:
@matheusd, remember to re-request review from reviewers when ready

matheusd commented on routing/missioncontrol_store.go:
```diff
@@ -268,8 +273,17 @@ func deserializeResult(k, v []byte) (*paymentResult, error) {
 	b.queueMx.Lock()
-	defer b.queueMx.Unlock()
 	b.queue.PushBack(rp)
+	signalRun := b.queue.Len() == 1
+	b.queueMx.Unlock()
+
+	// Signal run() that the queue is non-empty.
+	if signalRun {
+		select {
+		case <-b.done:
+		case b.queueChan <- struct{}{}:
+		}
+	}
```
`signalRun` is determined under the `queueMx` lock, which also gates the emptying of `b.queue`, therefore it will only ever be true if the `b.queue` was previously empty.

If this method (`AddResult`) is called in quick succession, one of two things can happen:

- Either `storeResults()` already started and emptied `b.queue`, in which case this will start a new batch and a new timer
- Or it has not started yet, and the new result will be put and processed in the outstanding batch

If this is called in quick succession with an empty `b.queue`, only the first one (under the `m.queueMx` lock) will send the signal.

matheusd commented:
@bitromortac sent commit https://github.com/lightningnetwork/lnd/pull/8549/commits/fb4367831fe8766f64e20288a9ab110304ce2f36 to clean up the mission control tests and address your comments.

Also rebased to address the conflict.

lightninglabs-deploy commented:
@bitromortac: review reminder
@matheusd, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@bitromortac: review reminder
@matheusd, remember to re-request review from reviewers when ready

ellemouton commented on routing/router_test.go:
```diff
@@ -136,6 +136,7 @@ func createTestCtxFromGraphInstanceAssumeValid(t *testing.T,
 		graphInstance.graphBackend, route.Vertex{}, mcConfig,
 	)
 	require.NoError(t, err, "failed to create missioncontrol")
+	drainMCStoreQueueChan(t, mc.store)
```
the addition of this miiight be worth its own commit. Not immediately clear why this is required

ellemouton commented on routing/missioncontrol_store.go:
```diff
@@ -339,12 +339,52 @@ func (b *missionControlStore) storeResults() error {
 
+	// Create a deduped list of new entries.
+	newKeys = make(map[string]*paymentResult, l.Len())
+	for e := l.Front(); e != nil; {
```
use `for e := l.Front(); e != nil; e = e.Next() {` then dont have to call e.Next manually?

ellemouton reviewed (commented):
Nice improvements! Thanks for this 🙏 

Main suggestion is about how to how to use `sync.Cond` instead of a channel for signalling & responding to new items.

Then more generally, some commit structure updates are needed: looks like some review comments were addressed in an additional commit rather than squashing those into the commit in question.

matheusd commented on routing/missioncontrol_store_test.go:
```diff
@@ -148,3 +150,242 @@ func TestMissionControlStore(t *testing.T) {
+	require.NoError(t, err)
+	store.run()
+	defer store.stop()
+	time.Sleep(flushInterval)
```
For the case where no updates happen, there is no way to know a flush happened. This is here to assert that, even after the flush interval elapses, no results are magically appearing.

The "correct" way to assert this would be to add a hook to be called after every flush, but that would only be useful for this specific test and would add an unnecessary (albeit very very small) overhead on production. And in any case, we don't even expect any results for the baseline: something would have to be very very wrong for there to appear a result here. So I took a shortcut.

matheusd replied:
(deleted the prior comment because I was looking at the wrong code)

matheusd replied:
This is here to assert that even after the first flush interval (i.e. the first potential flush), the results are correct (i.e. results have not been added or removed) after recreation.

The "correct" way to go about this would be to add a hook called after after every flush to determine when a flush happened and then perform the check. This would only be useful during testing and would increase complexity of the MC structure and add a (very very small) overhead on production, so I chose to take a shortcut here, since we're just asserting the baseline (that the results are the same).

Note that this (adding the hook and waiting on it) would mean the assertion would change during the PR: before my PR, the flush is continuous (so we'd have to assert the flush happened) and afterwards, the flush is only after a change (so we'd have to assert that no flush happened) - meaning the test would get more coupled to the implementation, also making it ever slightly more brittle.

If you feel this is the right approach here though, I'll make the appropriate changes.

ellemouton replied:
ok cool - makes sense - fine to leave as is I think! Just wanted to understand the reason 👍

matheusd commented on routing/missioncontrol_store_test.go:
```diff
@@ -148,3 +150,242 @@ func TestMissionControlStore(t *testing.T) {
+	// Store a result and check immediately. There still shouldn't be
+	// any results stored (flush interval has not elapsed).
+	store.AddResult(nextResult())
+	assertResults(0)
+
+	// Assert that eventually the result is stored after being flushed.
+	assertResults(1)
```
That is indeed correct (re: flake possibility), which is one of the reasons I used a largeish flush interval (relevant on CI).

matheusd commented on routing/missioncontrol_store_test.go:
```diff
@@ -148,3 +150,242 @@ func TestMissionControlStore(t *testing.T) {
+		nbResults int
+		doBench   func(b *testing.B, store *missionControlStore)
+	}{{
+		name:      "no additional results",
```
Modified.

matheusd commented on routing/missioncontrol_store_test.go:
```diff
@@ -148,3 +150,242 @@ func TestMissionControlStore(t *testing.T) {
+// works correctly.
+func TestMissionControlStoreFlushing(t *testing.T) {
+	// Set time zone explicitly to keep test deterministic.
+	time.Local = time.UTC
```
Rebased changes to first commit.

matheusd commented:
@ellemouton Thank you for the thorough review! I believe I addressed all of your points.

Also, rebased against current master.

Not sure what to do about the release notes, given the v0.18 RCs are already out (move to v0.19 or v0.18.1 or something else).

ellemouton commented:
thanks for the quick turn around! I'll check it out soon.

im also not sure re release notes.... cc @saubyk  for that :)

saubyk commented:
tagged for 18.1

saubyk commented:
> tagged for 18.1

On a second thought, 18.1 is already getting bloated and this doesn't seem urgent so moving to 0.19

matheusd commented:
Removed commit with type safety and moved release notes to v0.19.0.md.

lightninglabs-deploy commented:
@bitromortac: review reminder
@matheusd, remember to re-request review from reviewers when ready

ellemouton commented:
@matheusd  - note the failing linter check

bitromortac commented on routing/missioncontrol_store.go:
```diff
@@ -355,12 +355,51 @@ func (b *missionControlStore) storeResults() error {
 	var (
-		keys       *list.List
-		keysMap    map[string]struct{}
+		newKeys    map[string]*paymentResult
```
nit: this could be `map[string]struct{}`

bitromortac reviewed (approved):
LGTM :tada: (some nits and the linter fix left)

matheusd commented on routing/missioncontrol_store.go:
```diff
@@ -304,12 +306,25 @@ func (b *missionControlStore) run() {
+	// Return early when no new items will be updated.
+	if isEmpty {
+		return nil
+	}
```
Modified. 

Just as clarification, I did it in the original way to avoid adding the two mutex `Unlock()` calls. This is not an issue here because this section has very few lines, but I find that in early returns with mutexes involved, it's easy to make a mistake in the future, by adding an additional clause and forgetting to add the necessary Unlock call.

ellemouton commented on routing/missioncontrol_store_test.go:
```diff
@@ -131,20 +126,169 @@ func TestMissionControlStore(t *testing.T) {
+		tc := tc
+		name := fmt.Sprintf("%v additional results", tc)
+		b.Run(name, func(b *testing.B) {
+			h := newMCStoreTestHarness(b, testMaxRecords,
+				time.Second)
```
nit: formating

ellemouton replied:
unaddressed

ellemouton commented on routing/missioncontrol_store.go:
```diff
@@ -359,68 +359,98 @@ func (b *missionControlStore) storeResults() error {
+	for e := l.Front(); e != nil; e = e.Next() {
+		pr, ok := e.Value.(*paymentResult)
+		if !ok {
+			panic(fmt.Sprintf("wrong type %T (not *paymentResult)",
```
please dont use `panic` 🙏  return an error rather

matheusd replied:
Just to give my reasoning, I thought it would be ok in this situation because this is really a programmer error (no different than not checking for a nil argument on every function) and this was entirely avoidable by using the generic version of the lists (which was deemed as unnecessary in the review).

But in any case, I'll address it next.

ellemouton reviewed (changes requested):
almost there  - main thing is to please remove the use of `panic` 🙏

ellemouton commented on docs/release-notes/release-notes-0.19.0.md:
```diff
@@ -32,6 +32,9 @@
+
 # Technical and Architectural Updates
 ## BOLT Spec Updates
 ## Testing
```
missing your name in contributor list :)

ellemouton commented on routing/missioncontrol_store_test.go:
```diff
@@ -131,20 +126,169 @@ func TestMissionControlStore(t *testing.T) {
+	assertResults(testMaxRecords)
+}
+
+// BenchmarkMissionControlStoreFlushingNoWork benchmarks the periodic storage
```
unaddressed

ellemouton reviewed (approved):
LGTM once final nits have been addressed (note: I think you may have missed 2 from the last review)

matheusd commented:
Sorry for the misses. Updated.

lightninglabs-deploy commented:
@matheusd, remember to re-request review from reviewers when ready

ellemouton commented:
@saubyk  - happy for this to be merged? I think it is g2g

saubyk commented:
Tagged for 18.1. Release notes need update

ellemouton commented:
@matheusd  - thanks for your patience 🙏  I think we wanna move this to 0.18.1 instead of 0.19 so do you mind doing one last update and moving the release notes to the 0.18.1 doc pls?

matheusd commented:
[Updated](https://github.com/lightningnetwork/lnd/compare/e33d572bd222dfe1fd952adbaa8717011802cdd1..8f3dbdc7994ccfe94868d12b9d9168c3e33db526)

saubyk reviewed (approved):
Ack

ellemouton commented:
oh no @matheusd  - looks like there is a merge conflict 🙈  one more rebase pls 🙏  will then merge this asap so that this can be the last one!
